### ISSUE TITLE
[New Concept Exercise]: Anonymous Functions (lambdas) in Python

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3106]
This issue describes how to implement  the  `anonymous functions (lambdas)` in Python  **concept exercise** for the Python track.
The related **concept documents issue** can be found [here](https://github.com/exercism/python/issues/3105).

In addition, there was some previous discussion in the old concept exercise issue which you can find [here](https://github.com/exercism/python/issues/2357) and the closed PR, which you can find [here](https://github.com/exercism/python/pull/2589)

<br>

## ✅ Getting started

_If you have not yet created or contributed to a concept exercise, this issue will require some upfront reading to give you the needed background knowledge. Some good example exercises to look at in the repo:_

<details>
<summary>💡<b>Example Exercises</b>💡 (<em>click to expand</em>)</summary>
<br>

   1. [Little Sister's Vocabulary](https://github.com/exercism/python/tree/main/exercises/concept/little-sisters-vocab)
   2. [Meltdown Mitigation](https://github.com/exercism/python/tree/main/exercises/concept/meltdown-mitigation) 
   3. [Making the Grade](https://github.com/exercism/python/tree/main/exercises/concept/making-the-grade)
   4. [Ellen's Alien Game](https://github.com/exercism/python/tree/main/exercises/concept/ellens-alien-game)

</details>

We also recommend completing one or more of the concept exercises (_they're called "learning exercises"_) [on the website](https://exercism.org/tracks/python/concepts). 

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please go through the following documents:

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concept Exercises**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concept Exercises?](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Stories](https://exercism.org/docs/building/tracks/stories)

## 🎯 Goal

The goal of this concept exercise is to teach an understanding/use/creation of `anonymous functions (lambdas)`in Python.

<br>

## 💡Learning objectives

- Understand what an `anonymous function` is, and how to create one
   - The syntax of creating a  `lambda` 
   - Using different `function argument` flavors with `lambda`
- Understand the differences between `lambdas` and Pythons "regular" `functions`
- Understand what problems are solved by using a `lambda`
- The pitfalls of `lambdas`, and when to avoid them
- Using `lambdas` as `key functions` in other situations such as `sort()` ,  `sorted()`, `min()`,  and `max()`
- Applying arguments to a `lambda` via IIFE (_immediately invoked function expression_)
- Anti-patterns when using `lambdas` 

<br>

## 🤔 Concepts

- `anonymous-functions`
- `lambdas`
- `functions`, 
- `higher-order functions`
- `functions as arguments`
- `functions as returns`
- `nested funcitons`

<br>

## 🚫 Topics that are Out of scope

<br>
<details>
<summary><b>Concepts & Subjects that are Out of Scope</b> (<em>click to expand</em>)</summary>
<br>

- `comprehensions`
-  `lambdas` in `comprehensions`
- `comprehensions` in `lambdas`
- using a `decorator` on a `lambda`
- `functools` (_this will get its own exercise_)
- `generators`
- `map()`, `filter()`, and `reduce()` (_these will get their own exercise_)
- using an `assignment expression` or "walrus" operator (`:=`) in a `lambda`
-  assigning a variable to a `lambda`
-  using a `lambda` where a named function is recommended/better

</details>
<br>

## ↩️  Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  
<br>

<details>
<summary>Prereqs  (<em>click to expand</em>)</summary>

</details>

<br>

##  📚 Resources for Writing and Reference

<details>
<summary><b>Resources</b> (<em>click to expand</em>)</summary>
<br>

- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Python Docs Tutorial: Lambda Expressions](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions)
- [Functions as Objects in Python](https://medium.com/python-pandemonium/function-as-objects-in-python-d5215e6d1b0d)
- [Composing Programs: Higher-Order Functions](https://composingprograms.com/pages/16-higher-order-functions.html)
- [Learn by Example:  Python Lambda Function](https://www.learnbyexample.org/python-lambda-function/)
- [Real Python: How to Use Python Lambda Fuctions](https://realpython.com/python-lambda/)
- [Trey Hunner: Overusing Lambda expressions in Python](https://treyhunner.com/2018/09/stop-writing-lambda-expressions/)

</details>
<br>  

### Exercise Ideas & Stories

Should you need inspiration for an exercise story, you can find a collection [here](https://exercism.org/docs/building/tracks/stories).  You can also port an exercise from another track, **but please make sure to only to include tasks that actually make sense in Python** and that add value for a student.  Remove/replace/add tasks as needed to make the concept clear/workable.

<br>

##  📁 Exercise Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b> (<em>click to collapse</em>)</summary>
<table>
<td>
<br>

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This can summarize/paraphrase the linked concept documents if they have already been created (either the `about` or the `introduction`).   The summary does need to have enough information and examples  for the student to complete all the tasks outlined for this concept exercise.

* ### Exercise `instructions.md`

   For more information, see [`instructions.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsinstructionsmd)

   Instructions for an exercise usually center on a story that sets up the code challenge to be solved.  You can create your own story, or fork one from the ones listed [here](https://exercism.org/docs/building/tracks/stories).  Please make sure to give credit to the original authors if you use a story or fork an exercise.

* ### Exercise `Exemplar.py` Solution
  For more information, see [exemplar implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-exemplar-implementation).
  
  This file should not use syntax or datas structures not introduced in this exercise or in this exercise's prerequisites.  It will be used as an "ideal" solution for the challenge, so make sure it conforms to PEP8 and other formatting conventions, and **does not use single letter variable names**.  It should also include proper module and function-level docstrings.  However, it should **NOT** include typehinting or type aliases.

* ### `<Exercise>.py` (Stub) for Implementation
  For more information, see [stub implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-stub-implementation).
  
  This file should provide the expected function names imported for testing, and optionally TODO comments and or docstrings to aid the student in their implementation.  TODOs and docstrings are not required.

* ### `<Exercise>_Test.py` Files
  For more information, see [Tests](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-tests).
  Additionally, please note that Python associates exercise tasks to tests via a [Pytest Marker](https://docs.pytest.org/en/7.1.x/example/markers.html), and uses [`unittest subtests`](https://docs.python.org/3/library/unittest.html#subtests) as a form of test paramaterization.  See the test file for [`Little Sisters Vocab`](https://github.com/exercism/python/blob/main/exercises/concept/little-sisters-vocab/strings_test.py) for examples of how these techniques work.

* ### Exercise `Hints.md`

  For more information on writing hints see [`hints.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  - Hints should provide enough information to get most students "un-stuck" and moving toward a solution.  They should **not** provide a student with a direct solution.
  - You can refer to one or more of the resources linked in this issue above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### Exercise Metadata Files Under  `.meta/config.json`

  For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata-files)

   * `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
   * `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

<br>
</td>
</table>
</details>
<br>

## ♾️ Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [`config.json`](https://exercism.org/docs/building/tracks/config-json).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** :  `1e636fd3-6143-484e-a4fc-0ed6157fdfa1`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue    

<br>

## 🎶 Implementation Notes

- As a reminder, code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not strictly require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered either in the introduction to this exercise, or to one of its prerequisites.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

    - All asserts should contain a "user friendly" failure message (_these will display on the webiste to students, so be as clear as you can_). 
    - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
    - We also use `unittest.subtest` to parameterize test input where/when needed.
Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest-specific test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.

**Comment by DevanshRawat28671** [comment]
@BethanyG  I'd like to work on this

**Comment by BethanyG** [comment]
ghost -- I can't assign you since your GitHub profile has been deleted.  If you've created a new profile, please re-comment if you'd like to work on this issue, and I can assign you.  Many thanks!

**Comment by kana800** [comment]
hey @BethanyG, if @ghost isnt working on it; I can give this a try !

**Comment by BethanyG** [comment]
Hi @kana800  👋🏽 

_**I am so so sorry!**_. Your comment slipped through the notification cracks, and I am **just now** seeing it.  🤦🏽‍♀️   I do hope you are still interested, and if so ... this issue is **all yours**!  🎉  

 I will assign you, and just let me know if you still want to work on it/are no longer interested.  💙   

As a reminder, there was a previous draft created [here](https://github.com/exercism/python/pull/2589), but the author was using `lambdas` in a way that was not considered _pythonic_, so there is some discussion/reworking that probably needs to be done.  There are some additional links in the first part of the issue above as well.

**Comment by kana800** [comment]
@BethanyG hello ! I am sorry I just saw this; Sadly I am not able to work on it right away; I am gonna un-assign myself so that anyone else who is interested can work on this. 

Really sorry about the inconvenience.  

Thanks !

**Comment by safwansamsudeen** [comment]
@BethanyG, can I take this up? Thanks!

**Comment by BethanyG** [comment]
@safwansamsudeen  -- this exercise is closed to contribution.  Since the last contributors took it one, we've had a ton of other things (hints, approaches, mentor notes, earlier concept exercises, troubleshooting documents, additional practice exercises, earlier concept exercises ...) take higher priority.  Lambdas are also problematic in Python, and I want to think through exactly how we handle them for concept exercises.

I have left the issue here so that we don't lose the requirements and other notes.  I will flag it as paused.
